futures-util/src/compat/compat03as01.rs

```diff
@@ -108,19 +118,40 @@ where
     }
 }
 
+#[derive(Clone)]
 struct Current(task01::Task);
 
 impl Current {
     fn new() -> Current {
         Current(task01::current())
     }
 
-    fn as_waker(&self) -> task03::Waker {
-        // For simplicity reasons wrap the Waker into an Arc.
-        // We can optimize this again later on and reintroduce WakerLt<'a> which
-        // derefs to Waker, and where cloning it through RawWakerVTable returns
-        // an Arc version
-        ArcWake03::into_waker(Arc::new(Current(self.0.clone())))
+    fn as_waker(&self) -> WakerRef<'_> {
+        unsafe fn ptr_to_current<'a>(ptr: *const ()) -> &'a Current {
+            &*(ptr as *const Current)
+        }
+        fn current_to_ptr(current: &Current) -> *const () {
+            current as *const Current as *const ()
+        }
+
+        unsafe fn clone(ptr: *const ()) -> RawWaker {
+            // Lazily create the `Arc` only when the waker is actually cloned.
+            // FIXME: remove `transmute` when a `Waker` -> `RawWaker` conversion
+            // function is landed in `core`.
+            mem::transmute::<task03::Waker, RawWaker>(
```

Comment:
A method on ArcWake or the same module that returns a RawWaker seems cleaner than transmute.

The ideal solution might be to directly reference and reuse a refcounted thing within Task01. But after looking up the definition of that I don't think it's possible without the extra allocation and wrapping.